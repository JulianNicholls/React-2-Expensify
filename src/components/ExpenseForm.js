import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

class ExpenseForm extends React.Component {
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    pickerFocused: false,
    error: ''
  };

  changeDescription = e => {
    const description = e.target.value;

    this.setState(() => ({ description }));
  };

  changeNote = e => {
    const note = e.target.value;

    this.setState(() => ({ note }));
  };

  changeAmount = e => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
      this.setState(() => ({ amount }));
  };

  onDateChange = createdAt => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ pickerFocused: focused }));
  };

  addExpense = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'You must enter a description and an amount' }));
    } else {
      this.setState(() => ({ error: '' }));

      this.props.handleSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p className="form-error">{this.state.error}</p>}
        <form onSubmit={this.addExpense} className="expense-form">
          <input
            value={this.state.description}
            type="text"
            placeholder="Description"
            autoFocus={true}
            onChange={this.changeDescription}
          />
          <input
            value={this.state.amount}
            type="text"
            placeholder="Amount"
            onChange={this.changeAmount}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.pickerFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat={'ll'}
          />
          <textarea
            value={this.state.note}
            placeholder="Add an optional note for this expense"
            onChange={this.changeNote}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
