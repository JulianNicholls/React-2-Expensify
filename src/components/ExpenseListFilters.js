import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import * as actions from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    pickerFocused: null
  };

  onTextChange = e => this.props.setTextFilter(e.target.value);

  onSortChange = e =>
    e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = pickerFocused => this.setState(() => ({ pickerFocused }));

  render() {
    const {
      filters: { text, sortBy, startDate, endDate }
    } = this.props;

    return (
      <div>
        <input type="text" value={text} onChange={this.onTextChange} />
        <select value={sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={startDate}
          startDateId="sdid" // New requirement
          endDate={endDate}
          endDateId="edid" // New requirement
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.pickerFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          monthFormat={'MMM YY'}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

export default connect(
  mapStateToProps,
  actions
)(ExpenseListFilters);
