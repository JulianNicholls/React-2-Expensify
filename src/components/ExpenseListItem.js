import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import 'numeral/locales/en-gb';

numeral.locale('en-gb');

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div style={{ display: 'flex', alignItems: 'center', marginBottom: '7px' }}>
    <Link style={{ width: '60%' }} to={`/edit/${id}`}>
      <h4 style={{ margin: '0' }}>{description}</h4>
    </Link>
    <div style={{ width: '15%', textAlign: 'right' }}>
      {numeral(amount / 100.0).format('$0,0.00')}
    </div>
    <div style={{ width: '15%', paddingLeft: '2em' }}>
      {moment(createdAt).format('ll')}
    </div>
  </div>
);

export default ExpenseListItem;
