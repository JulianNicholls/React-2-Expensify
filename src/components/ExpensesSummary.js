import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import filteredExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expensesTotal';

import 'numeral/locales/en-gb';
numeral.locale('en-gb');

export const ExpensesSummary = ({ expensesTotal, expensesCount, filteredCount }) => {
  const expenseWord = filteredCount === 1 ? 'expense' : 'expenses';
  const ofWord = filteredCount < expensesCount ? ` (of ${expensesCount})` : '';

  if (filteredCount === 0)
    return (
      <h1>
        No {expenseWord}
        {ofWord} to display
      </h1>
    );

  return (
    <h1>
      Viewing {filteredCount} {expenseWord}
      {ofWord}, totalling {numeral(expensesTotal / 100).format('$0,0.00')}
    </h1>
  );
};

const mapStateToProps = state => {
  const filtered = filteredExpenses(state.expenses, state.filters);

  return {
    expensesTotal: expensesTotal(filtered),
    expensesCount: state.expenses.length,
    filteredCount: filtered.length
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
