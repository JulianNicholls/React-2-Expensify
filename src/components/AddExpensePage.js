import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export const AddExpensePage = ({ addExpense, history }) => (
  <div className="container">
    <h1>Add Expense</h1>
    <ExpenseForm
      handleSubmit={expense => {
        addExpense(expense);

        history.push('/');
      }}
    />
  </div>
);

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ addExpense }, dispatch);
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
