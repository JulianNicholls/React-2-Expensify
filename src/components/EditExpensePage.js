import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import * as actions from '../actions/expenses';

export class EditExpensePage extends React.Component {
  handleSubmit = updated => {
    this.props.editExpense(this.props.expense.id, updated);

    this.props.history.push('/');
  };

  handleRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);

    this.props.history.push('/');
  };

  render() {
    return (
      <div className="container">
        <h2>Edit Expense</h2>
        <ExpenseForm expense={this.props.expense} handleSubmit={this.handleSubmit} />
        <button onClick={this.handleRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(item => item.id === props.match.params.id)
  };
};

export default connect(
  mapStateToProps,
  actions
)(EditExpensePage);
