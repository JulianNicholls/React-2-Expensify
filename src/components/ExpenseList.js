import React from 'react';
import { connect } from 'react-redux';

import filteredExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = ({ expenses }) => (
  <div>{expenses.map(item => <ExpenseListItem key={item.id} {...item} />)}</div>
);

const mapStateToProps = state => ({
  expenses: filteredExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
