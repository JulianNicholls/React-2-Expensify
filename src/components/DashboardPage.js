import React from 'react';

import ExpenseList from './ExpenseList';

const DashboardPage = () => (
  <div className="container">
    <h1>Dashboard</h1>
    <ExpenseList />
  </div>
);

export default DashboardPage;
