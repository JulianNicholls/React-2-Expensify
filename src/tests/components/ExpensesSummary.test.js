import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';

describe('ExpensesSummary', () => {
  let defaultWrapper;

  beforeEach(() => {
    defaultWrapper = shallow(
      <ExpensesSummary expensesTotal={123245} filteredCount={4} expensesCount={4} />
    );
  });

  it('should render correctly', () => {
    expect(defaultWrapper).toMatchSnapshot();
  });

  it('should show the correct number of expenses and the correct total', () => {
    const text = defaultWrapper.find('h1').text();

    expect(text).toContain('4 expenses');
    expect(text).toContain('1,232.45');
  });

  it('should show when only some expenses are being displayed', () => {
    const wrapper = shallow(
      <ExpensesSummary expensesTotal={123245} filteredCount={3} expensesCount={4} />
    );

    const text = wrapper.find('h1').text();

    expect(text).toContain('3 expenses (of 4)');
    expect(text).toContain('1,232.45');
  });

  it('should show when only one expenses is being displayed', () => {
    const wrapper = shallow(
      <ExpensesSummary expensesTotal={123245} filteredCount={1} expensesCount={4} />
    );
    
    const text = wrapper.find('h1').text();

    expect(text).toContain('1 expense (of 4)');
    expect(text).toContain('1,232.45');
  });

  it('should show when all expenses have been filtered out', () => {
    const wrapper = shallow(
      <ExpensesSummary expensesTotal={0} filteredCount={0} expensesCount={4} />
    );
    const text = wrapper.find('h1').text();

    expect(text).toContain('No expenses (of 4) to display');
  });
});
