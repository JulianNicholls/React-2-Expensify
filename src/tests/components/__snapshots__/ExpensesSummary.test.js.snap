// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpensesSummary should render correctly 1`] = `
<h1>
  Viewing 
  4
   
  expenses
  , totalling 
  Â£1,232.45
</h1>
`;
